test/txstart-test.js

```diff
@@ -0,0 +1,225 @@
+const common = require('../lib/blockchain/common');
+const ownership = require('../lib/covenants/ownership');
+const VERIFY_NONE = common.flags.VERIFY_NONE;
+
+const node = new FullNode({
+  memory: true,
+  network: 'regtest',
+  plugins: [require('../lib/wallet/plugin')]
+});
+
+const RESET_TXSTART = node.network.txStart;
+
+// Ease up that mempool
+node.mempool.options.minRelay = 0;
+node.mempool.options.rejectAbsurdFees = false;
+node.miner.options.minWeight = 1000000;
+
+// Use a valid, working airdrop proof
+const FAUCET_PROOF_FILE = resolve(__dirname, 'data', 'faucet-proof.base64');
+const raw = Buffer.from(fs.readFileSync(FAUCET_PROOF_FILE, 'binary'), 'base64');
+const proof = AirdropProof.decode(raw);
+
+// We only need this for the fakeClaim
+const wallet = new MemWallet({network: 'regtest'});
+node.chain.on('connect', (entry, block) => {
+  wallet.addBlock(entry, block.txs);
+});
+wallet.getNameStatus = async (nameHash) => {
+  assert(Buffer.isBuffer(nameHash));
+  const height = node.chain.height + 1;
+  const state = await node.chain.getNextState();
+  const hardened = state.hasHardening();
+  return node.chain.db.getNameStatus(nameHash, height, hardened);
+};
+
+describe('txStart', function() {
+  let utxo, lastTX;
+
+  before(async () => {
+    node.network.txStart = 5;
+    await node.open();
+
+    // Start with one block for the fakeClaim
+    const block = await node.miner.mineBlock();
+    assert(await node.chain.add(block));
+  });
+
+  after(async () => {
+    await node.close();
+    node.network.txStart = RESET_TXSTART;
+  });
+
+  it('Should reject TX from mempool before txStart', async () => {
+    const tx = new TX({
+      inputs: [new Input()],
+      outputs: [new Output()]
+    });
+    tx.inputs[0].prevout.hash = Buffer.alloc(32, 0x01);
+
+    await assert.rejects(node.mempool.addTX(tx),
+      {reason: 'no-tx-allowed-yet'});
+  });
+
+  it('Should reject Claim from mempool before txStart', async () => {
+    const claim = await wallet.fakeClaim('cloudflare');
+
+    try {
+      ownership.ignore = true;
+      await assert.rejects(node.mempool.addClaim(claim),
+        {reason: 'no-tx-allowed-yet'});
+    } finally {
+      ownership.ignore = false;
+    }
+  });
+
+  it('Should reject Airdrop from mempool before txStart', async () => {
+    await assert.rejects(node.mempool.addAirdrop(proof),
+      {reason: 'no-tx-allowed-yet'});
+  });
+
+  it('Should reject block with >1 coinbase output before txStart', async () => {
+    const tx1 = new TX({
+      inputs: [new Input()],
+      outputs: [new Output(), new Output()]
+    });
+    tx1.locktime = node.chain.height + 1;
+
+    const block = new Block();
+    block.txs.push(tx1);
+    block.prevBlock = node.chain.tip.hash;
+    block.time = node.chain.tip.time + 1;
+    block.bits = await node.chain.getTarget(block.time, node.chain.tip);
+
+    await assert.rejects(node.chain.add(block, VERIFY_NONE),
+      {reason: 'no-tx-allowed-yet'});
```

Comment:
Another place where an assertion that the block was added to the invalids cache could be nice here.